src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -128,134 +112,239 @@ public static bool IsFlowSuppressed()
             return executionContext != null && executionContext.m_isFlowSuppressed;
         }
 
+        internal bool HasChangeNotifications => m_localChangeNotifications != null;
+
         public static void Run(ExecutionContext executionContext, ContextCallback callback, Object state)
         {
+            // Note: ExecutionContext.Run is an extremely hot function and used by every await, ThreadPool execution, etc.
             if (executionContext == null)
-                throw new InvalidOperationException(SR.InvalidOperation_NullContext);
+            {
+                ThrowNullContext();
+            }
 
-            Thread currentThread = Thread.CurrentThread;
-            ExecutionContextSwitcher ecsw = default(ExecutionContextSwitcher);
+            // enregistrer variables with 0 post-fix so they can be used in registers without EH forcing them to stack
+            // Capture references to Thread Contexts
+            Thread currentThread0 = Thread.CurrentThread;
+            Thread currentThread = currentThread0;
```

Comment:
This duplication relates to the above comment? This is ugly; we can't solve it by improving the JIT?

Reply:
Yes, we can improve the JIT.

For example [EH Write-Thru](../blob/master/Documentation/design-docs/eh-writethru.md) would allow the JIT to do better register allocation in the presence of EH.

Not sure when / if we will get to this.

Reply:
Added a note for it